### PR TITLE
Use Pango for text layout

### PR DESCRIPTION
**Pull request by asmaAL-Bahanta**
Use Pango for text layout to support languages that require complex text layout
such as Arabic, Hebrew, and Indic language.

The `cairo_show_text()` is part of cairo toy text API and should not be used in
serious applications, see:
http://cairographics.org/manual/cairo-text.html#cairo-show-text

### CODE REVIEW

**Comment by LebedevRI** [comment]
This will need a better description :)

**Comment by asmaAL-Bahanta** [comment]
I have modified some files so that it will render text using Pango instead of Cairo which does not support complex texts .

**Comment by LebedevRI** [comment]
Yes, that we can see, but the description should ideally say *why* that is needed, exactly.

**Comment by asmaAL-Bahanta** [comment]
It will add support for languages that requires complex text layout such as Arabic, Hebrew ...etc.

**Comment by asmaAL-Bahanta** [comment]
I modified the commit message and added more explanation.

File: src/iop/atrous.c [hunk 1]
```diff
@@ -1153,38 +1154,42 @@ static gboolean area_draw(GtkWidget *widget, cairo_t *crf, gpointer user_data)
   if(c->mouse_y > 0 || c->dragging)
   {
     // draw labels:
-    cairo_text_extents_t ext;
+    PangoLayout *layout;
+    PangoRectangle ink;
+    PangoFontDescription *desc = pango_font_description_from_string ("sans-serif bold");
```
Comment (houz) [inline src/iop/atrous.c]: You have to pass desc to `pango_font_description_free()` in the end.

File: src/iop/atrous.c [hunk 2]
```diff
@@ -1153,38 +1154,42 @@ static gboolean area_draw(GtkWidget *widget, cairo_t *crf, gpointer user_data)
...
+    layout = pango_cairo_create_layout(cr);
```
Comment (houz) [inline src/iop/atrous.c]: You have to pass layout to `g_object_unref()`.

File: src/iop/lowlight.c [hunk 3]
```diff
@@ -625,33 +625,42 @@ static gboolean lowlight_draw(GtkWidget *widget, cairo_t *crf, gpointer user_dat
   cairo_set_operator(cr, CAIRO_OPERATOR_SOURCE);
 
   // draw labels:
-  cairo_text_extents_t ext;
+  PangoLayout *layout;
+  PangoRectangle ink;
+  PangoFontDescription *desc = pango_font_description_from_string("sans-serif bold");
```
Comment (houz) [inline src/iop/lowlight.c]: You have to pass `desc` to `pango_font_description_free()` in the end.

File: src/iop/lowlight.c [hunk 4]
```diff
@@ -625,33 +625,42 @@ static gboolean lowlight_draw(GtkWidget *widget, cairo_t *crf, gpointer user_dat
...
+  pango_font_description_set_absolute_size(desc,(.06 * height) * PANGO_SCALE);
+  layout = pango_cairo_create_layout(cr);
```
Comment (houz) [inline src/iop/lowlight.c]: You have to pass `layout` to `g_object_unref()` in the end.

File: src/iop/lowlight.c [hunk 5]
```diff
@@ -625,33 +625,42 @@ static gboolean lowlight_draw(GtkWidget *widget, cairo_t *crf, gpointer user_dat
...
   cairo_set_font_size(cr, .06 * height);
 
-  cairo_text_extents(cr, _("dark"), &ext);
-  cairo_move_to(cr, .02 * width + ext.height, .5 * (height + ext.width));
+  pango_layout_set_text(layout, _("dark"), -1);
+  pango_layout_set_font_description(layout, desc);
+  pango_layout_get_pixel_extents(layout, &ink, NULL);
+  double y = .5 * (height + ink.width);
+  double x = .02 * width + ink.height;
+  cairo_move_to(cr, x- ink.height , y -ink.y );
```
Comment (houz) [inline src/iop/lowlight.c]: Why the two variables `x` and `y`? And why the `+ ink.height` in `x` if you subtract it again?

File: src/views/lighttable.c [hunk 6]
```diff
@@ -518,29 +518,37 @@ static int expose_filemanager(dt_view_t *self, cairo_t *cr, int32_t width, int32
     const float offy = height * 0.2f;
     const float offx = 60;
     const float at = 0.3f;
-    cairo_set_font_size(cr, fs);
+    PangoLayout *layout;
+    PangoFontDescription *desc = pango_font_description_from_string("sans-serif");
```
Comment (houz) [inline src/views/lighttable.c]: You have to pass `desc` to `pango_font_description_free()` in the end.

File: src/views/lighttable.c [hunk 7]
```diff
@@ -518,29 +518,37 @@ static int expose_filemanager(dt_view_t *self, cairo_t *cr, int32_t width, int32
...
+    pango_font_description_set_absolute_size(desc,(fs) * PANGO_SCALE);
     cairo_set_source_rgba(cr, .7, .7, .7, 1.0f);
-    cairo_select_font_face(cr, "sans-serif", CAIRO_FONT_SLANT_NORMAL, CAIRO_FONT_WEIGHT_NORMAL);
+    layout = pango_cairo_create_layout(cr);
```
Comment (houz) [inline src/views/lighttable.c]: You have to pass `layout` to `g_object_unref()` in the end.

**Comment by asmaAL-Bahanta** [comment]
I have addressed all the comments and made the changes. Still waiting for developers approval to change the remaining files which use `cairo_show_text()`

**Comment by houz** [comment]
Looks good, I will test it tomorrow. About the rest of the files, don't change it for things like the `G` in `paint.c`, there the easy cairo function is fine.

**Comment by khaledhosny** [comment]
> About the rest of the files, don't change it for things like the G in paint.c, there the easy cairo function is fine.

I think even for single letters using Caito toy text API is bad, as for example it does not do font fallback. My default UI font happens to lack any Latin letters, but it works just fine everywhere because font fallback is used, but when I opened Darktable for the first time I was greeted with [tofu](http://www.i18nguy.com/markup/i18n-glossary.html#tofu) boxes for what appeared to be some help messages. So I think it is better overall to just use Pango everywhere like GTK does and not have to worry about any of this.

**Comment by houz** [comment]
If you insist.

**Comment by houz** [comment]
Some problems:
* One of the lines on lighttable is starting at the wrong position now.
* The labels on left/right in equalizer are too far down and not aligned (they used to be right aligned).
* The numbers in tonecurve are too far down now.

**Comment by asmaAL-Bahanta** [comment]
I have made the changes and modified the positions.

**Comment by houz** [comment]
Thanks. Next time just push new commits please and never merge several into one.

**Pull request merged**